modules/core/src/main/java/org/apache/ignite/internal/processors/rollingupgrade/RollingUpgradeProcessor.java

```diff
@@ -0,0 +1,310 @@
+                metastorage.listen(ROLLING_UPGRADE_VERSIONS_KEY::equals, (key, oldVal, newVal) -> {
+                    verPair = (IgnitePair<IgniteProductVersion>)newVal;
+                });
+            }
+        });
+    }
+
+    /** {@inheritDoc} The joining node is stored to verify later whether it successfully connected to the ring or failed to join. */
+    @Override public @Nullable IgniteNodeValidationResult validateNode(ClusterNode node) {
+        synchronized (lock) {
+            lastJoiningNode = node;
+
+            lastJoiningNodeTimestamp = U.currentTimeMillis();
+        }
+
+        ClusterNode locNode = ctx.discovery().localNode();
+
+        String locBuildVer = locNode.attribute(ATTR_BUILD_VER);
+        String rmtBuildVer = node.attribute(ATTR_BUILD_VER);
+
+        IgniteProductVersion rmtVer = IgniteProductVersion.fromString(rmtBuildVer);
+
+        IgnitePair<IgniteProductVersion> pair = verPair;
+
+        if (pair == null)
+            pair = F.pair(IgniteProductVersion.fromString(locBuildVer), null);
+
+        IgniteProductVersion curVer = pair.get1();
+        IgniteProductVersion targetVer = pair.get2();
+
+        if (Objects.equals(rmtVer, curVer) || Objects.equals(rmtVer, targetVer))
+            return null;
+
+        String errMsg = "Remote node rejected due to incompatible version for cluster join.\n"
+            + "Remote node info:\n"
+            + "  - Version     : " + rmtBuildVer + "\n"
+            + "  - Addresses   : " + U.addressesAsString(node) + "\n"
+            + "  - Node ID     : " + node.id() + "\n"
+            + "Local node info:\n"
+            + "  - Version     : " + locBuildVer + "\n"
+            + "  - Addresses   : " + U.addressesAsString(locNode) + "\n"
+            + "  - Node ID     : " + locNode.id() + "\n"
+            + "Allowed versions for joining: " + curVer + (targetVer == null ? "" : ", " + targetVer);
+
+        LT.warn(log, errMsg);
+
+        if (log.isDebugEnabled())
+            log.debug(errMsg);
+
+        return new IgniteNodeValidationResult(node.id(), errMsg);
+    }
+
+    /**
+     * Enables rolling upgrade with specified target version.
+     * This method can only be called on coordinator node with {@link TcpDiscoverySpi}.
+     *
+     * @param target Target version.
+     * @throws IgniteCheckedException If:
+     *     <ul>
+     *         <li>The current and target versions are incompatible;</li>
+     *         <li>The local node is not a coordinator;</li>
+     *         <li>The discovery SPI is not {@link TcpDiscoverySpi};</li>
+     *         <li>The distributed metastorage is not ready;</li>
+     *         <li>{@link TcpDiscoveryNodesRing} was not provided.</li>
+     *     </ul>
+     */
+    public void enable(IgniteProductVersion target) throws IgniteCheckedException {
+        if (ring == null)
+            throw new IgniteCheckedException("TCP discovery nodes ring was not provided.");
```

Comment:
What admin should do with this message? And for "Meta storage is not ready"? 

Let's write a suggestion for an admin.